crates/bevy_math/src/sampling/mesh_sampling.rs

```diff
@@ -0,0 +1,66 @@
+//! Functionality related to random sampling from triangle meshes.
+
+use crate::{
+    primitives::{Measured2d, Triangle3d},
+    ShapeSample, Vec3,
+};
+use rand::Rng;
+use rand_distr::{Distribution, WeightedAliasIndex};
+use thiserror::Error;
+
+/// A [distribution] that caches data to allow fast sampling from a collection of triangles.
+/// Generally used through [`sample`] or [`sample_iter`].
+///
+/// [distribution]: Distribution
+/// [`sample`]: Distribution::sample
+/// [`sample_iter`]: Distribution::sample_iter
+///
+/// Example
+/// ```
+/// # use bevy_math::{Vec3, primitives::*};
+/// # use bevy_math::sampling::mesh_sampling::UniformMeshSampler;
+/// # use rand::{SeedableRng, rngs::StdRng, distributions::Distribution};
+/// let faces = Tetrahedron::default().faces();
+/// let sampler = UniformMeshSampler::try_new(faces).unwrap();
+/// let rng = StdRng::seed_from_u64(8765309);
+/// // 50 random points on the tetrahedron:
+/// let samples: Vec<Vec3> = sampler.sample_iter(rng).take(50).collect();
+/// ```
+pub struct UniformMeshSampler {
+    triangles: Vec<Triangle3d>,
+    face_distribution: WeightedAliasIndex<f32>,
+}
+
+impl Distribution<Vec3> for UniformMeshSampler {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec3 {
+        let face_index = self.face_distribution.sample(rng);
+        self.triangles[face_index].sample_interior(rng)
+    }
+}
+
+/// An error that indicates that a [`UniformMeshSampler`] could not be constructed
+/// because its input data had a total area of zero.
+#[derive(Debug, Error)]
+#[error("Failed to form distribution: provided triangles have zero area")]
+pub struct ZeroAreaMeshError;
+
+impl UniformMeshSampler {
+    /// Construct a new [`UniformMeshSampler`] from a list of triangles.
```

Comment:
Maybe a type-doc-link-thingy (I'm not well versed in the magics of docstrings) to Triangle3D would be good here?
```suggestion
    /// Construct a new [`UniformMeshSampler`] from a list of triangles [`Triangle3d`].
```